oeps/oep-0012.rst

```diff
@@ -0,0 +1,475 @@
+
+* A discussion that is shown within the "Teams" tab for each course team
+
+* In the future, there are thoughts of having program-wide discussions,
+  enterprise-wide discussions, and global discussions with a social group
+
+* There is also the concept of changing course announcements to be a new type of
+  discussion post
+
+In an ideal world, Open edx Discussions would be a separately-installable
+feature that would provide plugins for all of the functionality listed above. An
+Open edX operator could choose not to install edX discussions and choose to
+install different collaboration features instead.
+
+Currently the discussion board is a Django view built into edx-platform, but it
+should instead be made into a plugin. One option would be to create a new
+CourseDiscussionXBlock to display the "Course" tab, but as described above there
+are downsides to forcing this to be an XBlock. It would be cleaner if the edX
+discussion library provided a pluggable user interface that can be dropped into
+each of the contexts in which it is needed. The component would be provided
+with the context for authorization and to customize itself appropriately.
+
+An interesting question is how can such components could share course-scoped
+settings that the instructor specified in Studio. See `OEP-6 Context-scoped XBlock
+Fields`_ for the design of context-scoped fields in XBlock.
+
+Use Case 3: Instructor Dashboard
+--------------------------------
+
+The Instructor Dashboard provides a set of user interfaces to allow an
+instructor to manage the features of their course. Currently these user
+interfaces must be added directly to edx-platform, even if the feature
+itself lives in a Python library that has been separately installed.
+
+It would be preferable if any feature implemented as a Python library could
+contribute new sections to the Instructor Dashboard through plugins. The
+dashboard would show a list of the feature that can be managed, and then
+clicking on a feature would present the user with the pluggable interface.
+These interfaces could be made available for instructors in the LMS as
+today, or a new instructor portal could be implemented reusing the same
+blocks.
+
+Specification
+=============
+
+Requirements
+------------
+
+What are the requirements for a pluggable user interface?
+
+1. **It should render its content into a section of a page**
+
+   A pluggable user interface should be able to return the HTML, CSS and custom
+   JavaScript necessary to render its content. In addition, it should be able to
+   declare the JavaScript libraries that it depends upon.
+
+   Python and JavaScript APIs will be provided so that the consumer of a
+   component can render it. These APIs will ensure that the necessary libraries
+   have been loaded, and that the necessary CSS and JavaScript files have been
+   added to the page (without duplication).
+
+2. **It should be able to get the data that it needs to render its content**
```

Comment:
Thanks @robrap. 

> 1. Do you want to mention some part of this as part of the "Proposal" section, stating that this requirement will be fulfilled through standard Django mechanisms.

I'll beef up this section to make this more explicit.

> 2. Do we want to mention that there are or will be specific best practices and appropriate services to call inside edx-platform from one of these pluggable views, the details of which are out of scope of this OEP. The main point being to reiterate that someone should not feel free to do anything they please within the limits of Python and Django inside the platform.

I've added this to my upcoming draft.